packages/context/src/value-promise.ts

```diff
@@ -218,3 +217,55 @@ export function tryWithFinally<T>(
   }
   return result;
 }
+
+/**
+ * Resolve an iterator of source values into a result until the evaluator
+ * returns `true`
+ * @param source The iterator of source values
+ * @param resolver The resolve function that maps the source value to a result
+ * @param evaluator The evaluate function that decides when to stop
+ */
+export function resolveUntil<T, V>(
+  source: Iterator<T>,
+  resolver: (sourceVal: T) => ValueOrPromise<V | undefined>,
+  evaluator: (sourceVal: T, targetVal: V | undefined) => boolean,
+): ValueOrPromise<V | undefined> {
+  // Do iteration in loop for synchronous values to avoid stack overflow
+  while (true) {
+    const next = source.next();
+    if (next.done) return undefined; // End of the iterator
+    const sourceVal = next.value;
+    const valueOrPromise = resolver(sourceVal);
+    if (isPromiseLike(valueOrPromise)) {
+      return valueOrPromise.then(v => {
+        if (evaluator(sourceVal, v)) {
```

Comment:
Should we allow `evaluator` to return a `Promise<boolean>` too?

Reply:
We can add it if a use case comes up.